File: FraudGraph-Agent/backend/app/schemas/investigation.py
from __future__ import annotations

from datetime import UTC, datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


def utc_now() -> datetime:
    """Return the current timezone-aware UTC timestamp."""
    return datetime.now(UTC)


class InvestigationTriggerType(StrEnum):
    FRAUD_SIGNAL = "fraud_signal"
    CUSTOMER_REPORT = "customer_report"
    ANALYST_REQUEST = "analyst_request"


class InvestigationStatus(StrEnum):
    PENDING = "pending"
    INVESTIGATING = "investigating"
    AWAITING_EVIDENCE = "awaiting_evidence"
    AWAITING_APPROVAL = "awaiting_approval"
    ACTION_RECOMMENDED = "action_recommended"
    COMPLETED = "completed"
    ESCALATED = "escalated"
    FAILED = "failed"


class RiskLevel(StrEnum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"
    UNKNOWN = "unknown"


class InvestigationTrigger(BaseModel):
    model_config = ConfigDict(extra="forbid")

    trigger_type: InvestigationTriggerType

    transaction_id: str | None = Field(
        default=None,
        min_length=1,
    )

    customer_id: str | None = Field(
        default=None,
        min_length=1,
    )

    account_id: str | None = Field(
        default=None,
        min_length=1,
    )

    risk_score: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
    )

    reason: str | None = Field(
        default=None,
        min_length=1,
    )

    requested_by: str | None = Field(
        default=None,
        min_length=1,
    )

    metadata: dict[str, Any] = Field(
        default_factory=dict,
    )


class InvestigationCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    trigger: InvestigationTrigger

    priority: int = Field(
        default=0,
        ge=0,
        le=100,
    )

    requested_at: datetime = Field(
        default_factory=utc_now,
    )


class InvestigationProgress(BaseModel):
    model_config = ConfigDict(extra="forbid")

    step: int = Field(
        default=0,
        ge=0,
    )

    max_steps: int = Field(
        default=20,
        ge=1,
    )

    current_stage: str | None = None

    message: str | None = None

    completed_stages: list[str] = Field(
        default_factory=list,
    )


class InvestigationAssessment(BaseModel):
    model_config = ConfigDict(extra="forbid")

    risk_score: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
    )

    risk_level: RiskLevel = RiskLevel.UNKNOWN

    uncertainty: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
    )

    fraud_type: str | None = None

    confidence: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
    )

    rationale: str | None = None

    bank_risk_score: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
    )

    ml_fraud_probability: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
    )

    ml_used: bool = False

    ml_model_version: str | None = None

    graph_evidence_score: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
    )

    pattern_evidence_score: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
    )

    historical_case_score: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
    )

    trigger_type: str | None = None


class InvestigationResult(BaseModel):
    model_config = ConfigDict(extra="forbid")

    findings: list[str] = Field(
        default_factory=list,
    )

    recommendations: list[str] = Field(
        default_factory=list,
    )

    evidence_ids: list[str] = Field(
        default_factory=list,
    )

    related_case_ids: list[str] = Field(
        default_factory=list,
    )

    requires_additional_evidence: bool = False

    required_evidence: list[str] = Field(
        default_factory=list,
    )


class Investigation(BaseModel):
    model_config = ConfigDict(extra="forbid")

    investigation_id: str

    status: InvestigationStatus = InvestigationStatus.PENDING

    trigger: InvestigationTrigger

    progress: InvestigationProgress = Field(
        default_factory=InvestigationProgress,
    )

    assessment: InvestigationAssessment = Field(
        default_factory=InvestigationAssessment,
    )

    result: InvestigationResult = Field(
        default_factory=InvestigationResult,
    )

    case_id: str | None = None

    created_at: datetime = Field(
        default_factory=utc_now,
    )

    updated_at: datetime = Field(
        default_factory=utc_now,
    )

    completed_at: datetime | None = None

File: FraudGraph-Agent/frontend/types/investigation.ts
export type InvestigationStatus =
  | "pending"
  | "running"
  | "investigating"
  | "awaiting_evidence"
  | "awaiting_approval"
  | "completed"
  | "failed"
  | "cancelled";

export type InvestigationPriority =
  | "low"
  | "medium"
  | "high"
  | "critical";

export type InvestigationStage =
  | "initialization"
  | "data_collection"
  | "graph_analysis"
  | "risk_analysis"
  | "evidence_analysis"
  | "pattern_detection"
  | "decision"
  | "action"
  | "completed";

export interface InvestigationProgress {
  stage: InvestigationStage;
  progress: number;
  message: string | null;
  started_at: string | null;
  updated_at: string | null;
}

export interface InvestigationAgent {
  agent_id: string;
  name: string;
  type: string;
  status: string;
  message: string | null;
  started_at: string | null;
  completed_at: string | null;
}

export interface Investigation {
  investigation_id: string;

  case_id: string | null;

  transaction_id: string | null;

  customer_id: string | null;

  account_id: string | null;

  title: string;

  description: string | null;

  status: InvestigationStatus;

  priority: InvestigationPriority;

  risk_score: number | null;

  risk_level: string | null;

  progress: InvestigationProgress | null;

  agents: InvestigationAgent[];

  evidence_ids: string[];

  finding_ids: string[];

  decision_ids: string[];

  action_ids: string[];

  created_at: string;

  started_at: string | null;

  updated_at: string;

  completed_at: string | null;

  error: string | null;
}

export interface InvestigationCreate {
  case_id?: string | null;

  transaction_id?: string | null;

  customer_id?: string | null;

  account_id?: string | null;

  title: string;

  description?: string | null;

  priority?: InvestigationPriority;
}

export interface InvestigationUpdate {
  status?: InvestigationStatus;

  priority?: InvestigationPriority;

  title?: string;

  description?: string | null;

  risk_score?: number | null;

  risk_level?: string | null;
}

export interface InvestigationListResponse {
  items: Investigation[];

  total?: number;

  page?: number;

  page_size?: number;

  has_next?: boolean;
}

export interface InvestigationEvent {
  event_id?: string;

  investigation_id: string;

  type: string;

  timestamp: string;

  message?: string | null;

  data?: Record<string, unknown>;
}